Salmon - Classify Reads
---------------------------

Find closest homologue sequence for each sequenced fragment

**Location**

- *Filepath:* <SnakeLines_dir>/rules/paired_end/classification/transcripts/salmon.snake
- *Rule name:* salmon__classify_reads

**Input(s):**

- *r1:* Left side of sequenced fragments in gzipped fastq format
- *r2:* Right side of sequenced fragments in gzipped fastq format
- *index:* One of index files generated by Salmon

**Output(s):**

- *counts:* TSV file with estimated taxonomic counts for each transcript

Salmon - Create Reference Index
-----------------------------------

Transform transcriptomic sequences into Salmon index for faster classification

**Location**

- *Filepath:* <SnakeLines_dir>/rules/paired_end/classification/transcripts/salmon.snake
- *Rule name:* salmon__create_reference_index

**Input(s):**

- *fasta:* Genomic reference sequences of transcripts

**Output(s):**

- *index:* One of index files generated by Salmon

Salmon - Prepare For Krona
------------------------------

Convert Salmon classification files into standardised format suitable for generation of Krona reports

**Location**

- *Filepath:* <SnakeLines_dir>/rules/paired_end/classification/transcripts/salmon.snake
- *Rule name:* salmon__prepare_for_krona

**Input(s):**

- *counts:* Read counts from Salmon

**Output(s):**

- *krona:* Tabular format suitable for Krona report generation

