Bismark - Methylation Extractor
-----------------------------------

Reads in a bisulfite read alignment results file produced by the Bismark bisulfite mapper and extracts the
methylation information for individual cytosines

**Location**

- *Filepath:* <SnakeLines_dir>/rules/mapping/report/methylation/bismark.snake
- *Rule name:* bismark__methylation_extractor

**Input(s):**

- *bam:* mapped reads with Bismark in BAM format
- *ct_index:* CT reference index (created by rule bismark__prepare_index)
- *ga_index:* GA reference index (created by rule bismark__prepare_index)
- *ref:* reference genome, i.e. reference/{reference}/{reference}.fa

**Output(s):**

- *mbias_report:* report generated by the bismark_methylation_extractor tool
- *splitting_report:* report generated by the bismark_methylation_extractor tool

Bismark - Summary Report
----------------------------

Aggregate results from several methylation statistic files generated by Bismark into single summary HTML report

**Location**

- *Filepath:* <SnakeLines_dir>/rules/mapping/report/methylation/bismark.snake
- *Rule name:* bismark__summary_report

**Input(s):**

- *alignment_report:* Report generated with Bismark together with read alignment
- *splitting_report:* Report generated by the bismark_methylation_extractor tool
- *mbias_report:* Report generated by the bismark_methylation_extractor tool

**Output(s):**

- *html:* Aggregated quality report of discovered methylation

