configfile: srcdir('config.yaml')

pipeline = glob_wildcards('reads/original/{sample, %s}_R1.fastq.gz' % config['sample_pattern'])

include: config['snakelines_dir'] + '/rules/preprocess/report/fastqc.snake'
include: config['snakelines_dir'] + '/rules/preprocess/trim_reads/trimmomatic.snake'
include: config['snakelines_dir'] + '/rules/preprocess/filtering/bowtie2.snake'
include: config['snakelines_dir'] + '/rules/preprocess/join_reads/pear.snake'

include: config['snakelines_dir'] + '/rules/metagenomics/classification/rdp.snake'
include: config['snakelines_dir'] + '/rules/metagenomics/report/krona.snake'
include: config['snakelines_dir'] + '/rules/metagenomics/report/tsv_summary.snake'

genes = list(config['gene_pattern'].keys())

rule process:
    input:
        'reads/original/stats/summary.html',
        'reads/trimmed/stats/summary.html',
        'reads/filtered/stats/summary.html',
        'reads/joined/stats/summary.html',
        'reads/joined/stats-joined/summary.html',
        expand('rdp/{gene}/report/krona/krona.html', gene=genes),
        expand('rdp/{gene}/report/tsv/summary.tsv', gene=genes)