rule call_germline:
    input:
        bam = '{mapper}/{reference}/{sample}.dup.realign.bam',
        bai = '{mapper}/{reference}/{sample}.dup.realign.bam.bai',
        bed = 'reference/{reference}/annotation/{panel}/regions.interval_list',
        reffasta = 'reference/{reference}/{reference}.fa',
        reffaidx = 'reference/{reference}/{reference}.fa.fai'
    output:
        '{mapper}/{reference}/gatk/{sample}.{panel}.germline.vcf'
    log:
        '{mapper}/{reference}/gatk/log/{sample}.{panel}.germline.log'
    shell:
        '''
        gatk HaplotypeCaller \
            -R {input.reffasta} \
            -I {input.bam} \
            -L {input.bed} \
            -o {output} \
        > {log}
        '''
