THREADS = 16

rule clark:
    input:
        r1 = 'reads/{type}/{sample}_R1.fastq',
        r2 = 'reads/{type}/{sample}_R2.fastq'
    output:
        'clark/{type}/classification/{sample}/reads.csv'
    params:
        prefix='clark/{type}/classification/{sample}/reads',
        dir='clark/{type}/classification/{sample}'
    threads:
        THREADS
    shell:
        '''
        mkdir -p {params.dir}

        clark \
            -k 31 \
            -T /data/genome/metagenome/clark/17-01-18/targets.txt \
            -D /data/genome/metagenome/clark/17-01-18 \
            -P {input.r1} \
               {input.r2} \
            -R {params.prefix} \
            -m 1 \
            -n {threads}
        '''

rule clark_summary:
    input:
        'clark/{type}/classification/{sample}/reads.csv'
    output:
        'clark/{type}/classification/{sample}/summary.csv'
    shell:
        '''
        /usr/local/tools/clark-1.2.3/estimate_abundance.sh \
            -F {input} \
            -D /data/genome/metagenome/clark/17-01-18 \
        > {output}
        '''

rule krona_all_txt:
    input:
        'clark/{type}/classification/{sample}/summary.csv'
    output:
        'clark/{type}/report/krona/individual/{sample}.krn'
    shell:
        '''
        awk \
            -F, \
            '!/UNKNOWN/ {{ \
                gsub(";", "\t", $3); \
                print $4 "\t" $3 "\t" $1\
            }}' \
            {input} \
        | tail \
            -n +2 \
        > {output}
        '''

