configfile: srcdir('config.yaml')

pipeline = glob_wildcards('reads/original/{sample,%s}_R1.fastq.gz' % config['sample_pattern'])

include: config['snakelines_dir'] + '/rules/preprocess/report/fastqc.snake'
include: config['snakelines_dir'] + '/rules/preprocess/trim_reads/trimmomatic.snake'
include: config['snakelines_dir'] + '/rules/mapping/mappers/bowtie2.snake'
include: config['snakelines_dir'] + '/rules/mapping/postprocess/bam_index.snake'
include: config['snakelines_dir'] + '/rules/mapping/postprocess/read_group/custom.snake'
include: config['snakelines_dir'] + '/rules/mapping/postprocess/deduplication/picard.snake'
include: config['snakelines_dir'] + '/rules/mapping/postprocess/realignment/gatk.snake'
include: config['snakelines_dir'] + '/rules/mapping/report/qualimap.snake'
include: config['snakelines_dir'] + '/rules/variant/callers/vardict.snake'

reference = config['mapping']['reference']
panel = config['variant']['panel']

rule process:
    input:
        # Fastq quality control reports
        'reads/original/stats/summary.html',
        'reads/trimmed/stats/summary.html',

        # BAM files
        expand('bowtie2/{reference}/realign/{sample}.bam',
                reference=reference, sample=pipeline.samples),

        # Qualimap reports
        'bowtie2/{reference}/stats-{panel}/summary.tsv'.format(
                reference=reference, panel=panel),

        # Variant files
        expand('bowtie2/{reference}/vardict-{panel}/{sample}.germline.vcf',
                reference=reference, sample=pipeline.samples, panel=panel),

