configfile: srcdir('config.yaml')

read_type = config['subsample']['read_type']
comparisons = config['lefse']['comparisons']
gene = config['metaxa2']['gene']

pipeline = glob_wildcards('reads/%s/{sample, %s}_R1.fastq.gz' % (read_type, config['sample_pattern']))

include: config['snakelines_dir'] + '/rules/preprocess/subsample/head.snake'
include: config['snakelines_dir'] + '/rules/preprocess/report/fastqc.snake'
include: config['snakelines_dir'] + '/rules/metagenomics/classification/metaxa2.snake'
include: config['snakelines_dir'] + '/rules/metagenomics/report/krona.snake'
include: config['snakelines_dir'] + '/rules/metagenomics/report/tsv_summary.snake'
include: config['snakelines_dir'] + '/rules/metagenomics/comparison/lefse.snake'

rule process:
    input:
        # Summary tables
        'metaxa2/%s/report/krona/krona.html' % gene,
        'metaxa2/%s/report/tsv/summary.tsv' % gene,
        'metaxa2/%s/report/tsv/summary.xlsx' % gene,
        'metaxa2/%s/report/tsv/subtaxes.tsv' % gene,
        'metaxa2/%s/report/tsv/subtaxes.xlsx' % gene,
        'metaxa2/%s/report/tsv/subtaxes.norm.xlsx' % gene,
        'metaxa2/%s/report/tsv/subtaxes.norm.tsv' % gene,

        # Lefse comparison
        expand('metaxa2/{gene}/report/lefse/{comparison}/significant-taxes.pdf', comparison=comparisons, gene=gene),
        expand('metaxa2/{gene}/report/lefse/{comparison}/treeview.pdf',          comparison=comparisons, gene=gene),
        expand('metaxa2/{gene}/report/lefse/{comparison}/comparison.html',       comparison=comparisons, gene=gene)

