configfile: srcdir('config.yaml')

pipeline = glob_wildcards('reads/original/{sample,%s}_R1.fastq.gz' % config['sample_pattern'])

include: config['snakelines_dir'] + '/rules/preprocess/report/fastqc.snake'
include: config['snakelines_dir'] + '/rules/preprocess/trim_reads/trimmomatic.snake'
include: config['snakelines_dir'] + '/rules/preprocess/deduplication/fastuniq.snake'
include: config['snakelines_dir'] + '/rules/mapping/mappers/bowtie2.snake'
include: config['snakelines_dir'] + '/rules/mapping/postprocess/bam_index.snake'
include: config['snakelines_dir'] + '/rules/mapping/report/qualimap.snake'
include: config['snakelines_dir'] + '/rules/variant/callers/freebayes.snake'

reference = config['mapping']['reference']

rule process:
    input:
        # Preprocess
        'reads/original/stats/summary.html',
        'reads/trimmed/stats/summary.html',

        # Mapping
        expand('bowtie2/{reference}/{sample}.bam'    ,
               reference=reference, sample=pipeline.samples),

        expand('bowtie2/{reference}/{sample}.bam.bai',
               reference=reference, sample=pipeline.samples),

        'bowtie2/%s/stats-wgs/summary.tsv' % reference,

        # Variant calling
        expand('bowtie2/{reference}/freebayes/{sample}.population.vcf',
               reference=reference, sample=pipeline.samples),