read_type = config['mapping']['read_type']

rule map_reads:
    input:
        r1 = 'reads/%s/{sample}_R1.fastq.gz' % read_type,
        r2 = 'reads/%s/{sample}_R2.fastq.gz' % read_type,
        index = 'reference/{reference}/bowtie2_index/{reference}.1.bt2',
        reffasta = 'reference/{reference}/{reference}.fa'
    output:
        bam = 'bowtie2/{reference}/{sample}.bam',
    log:
        'bowtie2/{reference}/log/{sample}.map'
    params:
        index = 'reference/{reference}/bowtie2_index/{reference}',
        tmpdir = 'bowtie2/{reference}/tmp/',
        additional = config['mapping']['params']
    threads:
        int(config['threads'])
    shell:
        '''
        bowtie2 \
            -x {params.index} \
            -1 {input.r1} \
            -2 {input.r2} \
            --threads {threads} \
            {params.additional} \
        2> {log} \
        | samtools view \
            -b \
            --reference {input.reffasta} \
            --threads {threads} \
            - \
        | samtools sort \
            --threads {threads} \
            -o {output.bam} \
            -T {params.tmpdir} \
            -m 5G \
            - \
        '''

rule bowtie2_index:
    input:
        'reference/{genome}/{sequence}.fa'
    output:
        protected('reference/{genome}/bowtie2_index/{sequence}.1.bt2'),
        protected('reference/{genome}/bowtie2_index/{sequence}.2.bt2'),
        protected('reference/{genome}/bowtie2_index/{sequence}.3.bt2'),
        protected('reference/{genome}/bowtie2_index/{sequence}.4.bt2'),
        protected('reference/{genome}/bowtie2_index/{sequence}.rev.1.bt2'),
        protected('reference/{genome}/bowtie2_index/{sequence}.rev.2.bt2')
    params:
        index = 'reference/{genome}/bowtie2_index/{sequence}'
    log:
        'reference/{genome}/log/bowtie2.log'
    threads:
        int(config['threads'])
    shell:
        '''
        bowtie2-build \
            --threads {threads} \
            {input} \
            {params.index} \
        > {log}
        '''

