rule spades__assemble_reads_into_contigs
----------------------------------------
located in: '<SnakeLines_dir>/rules/assembly/assembler/spades.snake'

Assemble preprocessed reads into the larger genomic sequences, contigs. Also generate contig overlap graphs and
initial scaffolds. Use 'spades' program.

Input(s):
	r1: gzipped fastq file with left reads
	r2: gzipped fastq file with right reads
Output(s):
	fastq: assembly graph .fastq file
	gfa: assembly graph .gfa file ????
	contigs: contigs in .fa file
	scaffolds: scaffolds in .fa file
Param(s):
	outdir: output directory (do not change)
	contigs: contig file generated by spades (do not change)
	scaffolds: scaffolds file generated by spades (do not change)
	mode: mode of operation of spades (do not change), extracted from config file
	careful: whether to use --careful parameter for spades (do not change), extracted from config file

rule unicycler__assemble_reads_into_contigs
-------------------------------------------
located in: '<SnakeLines_dir>/rules/assembly/assembler/unicycler.snake'

Assemble preprocessed reads into the larger genomic sequences, contigs. Also generate contig overlap graphs and
initial scaffolds. Use 'spades' program.

Input(s):
	r1: gzipped fastq file with left reads
	r2: gzipped fastq file with right reads
Output(s):
	gfa: assembly graph .gfa file ????
	contigs: contigs in .fa file
Param(s):
	outdir: output directory (do not change)
	contigs: contig file generated by spades (do not change)
	gfa: assembly graph .gfa file ????

