configfile: srcdir('config.yaml')

pipeline = glob_wildcards('reads/original/{sample, %s}_R1.fastq.gz' % config['sample_pattern'])

include: config['snakelines_dir'] + '/rules/preprocess/report/fastqc.snake'
include: config['snakelines_dir'] + '/rules/preprocess/trim_reads/trimmomatic.snake'
include: config['snakelines_dir'] + '/rules/preprocess/deduplication/fastuniq.snake'
include: config['snakelines_dir'] + '/rules/metagenomics/classification/emirge.snake'
include: config['snakelines_dir'] + '/rules/metagenomics/report/krona.snake'
include: config['snakelines_dir'] + '/rules/metagenomics/report/tsv_summary.snake'

references = list(config['gene_pattern'].keys())

rule process:
    input:
        'reads/original/stats/summary.html',
        'reads/trimmed/stats/summary.html',
        expand('emirge/{reference}/report/krona/krona.html', reference=references),
        expand('emirge/{reference}/report/tsv/summary.tsv', reference=references),
        expand('emirge/{reference}/report/tsv/summary.xlsx', reference=references),
        expand('emirge/{reference}/report/tsv/subtaxes.tsv', reference=references),
        expand('emirge/{reference}/report/tsv/subtaxes.xlsx', reference=references),
        expand('emirge/{reference}/report/tsv/subtaxes.norm.tsv', reference=references),
        expand('emirge/{reference}/report/tsv/subtaxes.norm.xlsx', reference=references)