rule virfinder__identify_viral_sequences:
    """
    Method assigns each contig p-value that it originates from viral genome.
    :input fasta: Genomics sequences to be analysed in fasta format
    :output pvalues: Probabilities that genomic sequence originates from a virus
    :params wrapper_script: Wrapper script to call Virfinder
    """
    input:
        fasta   = '{indir}/{name}.fa'
    output:
        pvalues = '{indir}/virfinder/{name}.pvalues.tsv'
    log:
        out     = '{indir}/virfinder/log/{name}.log',
        err     = '{indir}/virfinder/log/{name}.err'
    params:
        wrapper_script = srcdir('virfinder/virfinder_wrapper.R')
    conda:
        config['snakelines_dir'] + '/enviroments/others_env.yaml'
    shell:
        """
        Rscript {params.wrapper_script} \
            {input.fasta} \
            {output.pvalues} \
         > {log.out} \
        2> {log.err}
        """
