rule gatk__fix_vcf_header:
    '''
    add missing sequence dictionary to vcf header
    this job also generates vcf index (.idx)
    '''
    input:
        vcf = 'variant/{reference}-wgs/original/{sample}.vardict.vcf',
        fasta = 'reference/{reference}/{reference}.fa'
    output:
        vcf = 'variant/{reference}-wgs/original/{sample}.vcf',
        vcf_index = 'variant/{reference}-wgs/original/{sample}.vcf.idx',
    log:
        out = 'variant/{reference}-wgs/original/log/{sample}.out',
        err = 'variant/{reference}-wgs/original/log/{sample}.err',
    shell:
        '''
        gatk4 SelectVariants \
            -V {input.vcf} \
            -R {input.fasta} \
            -O {output.vcf} \
            1>{log.out} \
            2>{log.err}
        '''

rule tabix__index_vcf:
    input:
        '{path}/{filename}.vcf.gz'
    output:
        '{path}/{filename}.vcf.gz.tbi'
    log:
        out = '{path}/log/{filename}_tabix.out',
        err = '{path}/log/{filename}_tabix.err'
    shell:
        '''
        tabix -p vcf {input} 1>{log.out} 2>{log.err}
        '''

rule picard__bed_to_interval_list:
    input:
        bed      = 'reference/{reference}/annotation/wgs/regions.bed',
        seq_dict = 'reference/{reference}/{reference}.dict'
    output:
        intervals = 'variant/{reference}-wgs/original/interval_list/{sample}.interval_list'
    log:
        out = 'variant/{reference}-wgs/original/interval_list/log/{sample}.out',
        err = 'variant/{reference}-wgs/original/interval_list/log/{sample}.err'
    shell:
        '''
        picard BedToIntervalList \
            INPUT={input.bed} \
            SEQUENCE_DICTIONARY={input.seq_dict} \
            OUTPUT={output.intervals} \
            SORT=true \
            UNIQUE=true \
            1> {log.out} \
            2> {log.err}
        '''

rule gatk__collect_variant_calling_metrics:
    input:
        vcf = 'variant/{reference}-{panel}/original/{sample}.vcf',
        vcf_index = 'variant/{reference}-{panel}/original/{sample}.vcf.idx',

        dbsnp = 'reference/{reference}/dbsnp/dbsnp.vcf.gz',
        dbsnp_index = 'reference/{reference}/dbsnp/dbsnp.vcf.gz.tbi',

        intervals = 'variant/{reference}-{panel}/original/interval_list/{sample}.interval_list'
    output:
        filename = 'variant/{reference}-{panel}/original/calling_metrics/{sample}.variant_calling_summary_metrics'
    log:
        out = 'variant/{reference}-{panel}/original/calling_metrics/log/{sample}.out',
        err = 'variant/{reference}-{panel}/original/calling_metrics/log/{sample}.err'
    shell:
        '''
        gatk4 CollectVariantCallingMetrics \
            --INPUT {input.vcf} \
            --DBSNP {input.dbsnp} \
            --OUTPUT variant/{wildcards.reference}-{wildcards.panel}/original/calling_metrics/{wildcards.sample} \
            --THREAD_COUNT {threads} \
            --TARGET_INTERVALS {input.intervals} \
            1> {log.out} \
            2> {log.err}
        '''