include: srcdir('indices/bowtie2_index.snake')

rule bowtie2__map_reads_to_reference:
    '''
    For input preprocessed reads bowtie2 finds the most similar genomic region in the provided reference genome.
    '''
    input:
        r1    = 'reads/%s/{sample}_R1.fastq.gz' % pipeline.preprocessed_read_type,
        r2    = 'reads/%s/{sample}_R2.fastq.gz' % pipeline.preprocessed_read_type,
        index = 'reference/{reference}/bowtie2_index/{reference}.1.bt2',
        ref   = 'reference/{reference}/{reference}.fa'
    output:
        bam   = 'mapping/{reference}/original/{sample}.bam'
    log:
        err   = 'mapping/{reference}/log/{sample}.bowtie2.map_stats'
    params:
        index = 'reference/{reference}/bowtie2_index/{reference}',
        additional = tool_config.get('params', ''),
        concordant = '--no-mixed --no-discordant --no-unal' if tool_config.get('only_concordant', False) else ''
    threads:
        int(config['threads'])
    shell:
        '''
        bowtie2 \
            -x {params.index} \
            -1 {input.r1} \
            -2 {input.r2} \
            --threads {threads} \
            {params.concordant} \
            {params.additional} \
        2> {log.err} \
        | samtools view -bS - \
        > {output.bam}
        '''
#
#rule bowtie2_remap_reads_to_assembly:
#    input:
#        r1 = 'reads/%s/{sample}_R1.fastq.gz' % config['mapping']['read_type'],
#        r2 = 'reads/%s/{sample}_R2.fastq.gz' % config['mapping']['read_type'],
#        index = '{fapath}/{sample}/bowtie2_index/{contigs}.1.bt2',
#        fasta = '{fapath}/{sample}/{contigs}.fa'
#    output:
#        bam = '{fapath}/{sample}/mapping/{contigs}.bam'
#    log:
#        '{fapath}/{sample}/mapping/log/{sample}.map'
#    params:
#        index = '{fapath}/{sample}/bowtie2_index/{contigs}',
#        tmpdir = '{fapath}/{sample}/tmp/',
#        additional = config['mapping']['params'] if 'params' in config['mapping'] else ''
#    threads:
#        int(config['threads'])
#    shell:
#        '''
#        bowtie2 \
#            -x {params.index} \
#            -1 {input.r1} \
#            -2 {input.r2} \
#            --threads {threads} \
#            {params.additional} -t \
#        2> {log} \
#        | samtools view \
#            -b \
#            --threads {threads} \
#            - \
#        | samtools sort \
#            --threads {threads} \
#            -o {output.bam} \
#            -T {params.tmpdir} \
#            -m 5G \
#            - \
#        '''
