configfile: srcdir('config.yaml')

pipeline = glob_wildcards('reads/original/{sample,%s}_R1.fastq.gz' % config['sample_pattern'])

include: config['snake_dir'] + '/preprocess/report/fastqc.snake'
include: config['snake_dir'] + '/preprocess/trim_reads/trimmomatic.snake'
include: config['snake_dir'] + '/preprocess/deduplication/fastuniq.snake'
include: config['snake_dir'] + '/assembly/assemblers/trinity.snake'
include: config['snake_dir'] + '/classification/classifiers/blast.snake'
include: config['snake_dir'] + '/classification/report/fasta_summary.snake'
include: config['snake_dir'] + '/transcript/quantifiers/kallisto.snake'
include: config['snake_dir'] + '/transcript/report/abundances.snake'

design = config['trinity']['design']

rule process:
    input:
        # FastQC summaries
        'reads/original/stats/summary.html',
        'reads/trimmed/stats/summary.html',
        'reads/deduplicated/stats/summary.html',

        # Contigs
        expand('trinity-{design}/contigs.fa',
                design=design),

        # Annotations for contigs
        expand('trinity-{design}/blast/contigs.html',
                design=design) if config['blast']['enabled'] else [],

        # Read quantification
        expand('trinity-{design}/kallisto/abundance.tsv',
                design=design),

        # HTML report
        expand('trinity-{design}/kallisto/contigs.html',
                design=design)